File: Data Structures & Algorithms/find-in-mountain-array/submission-3.py
class Solution:
    def findInMountainArray(self, target: int, mountainArr: 'MountainArray') -> int:
        
        def search1(i,j):
            while i<=j:
                mid = (i+j)//2
                if mountainArr.get(mid) < target:
                    i = mid+1
                elif mountainArr.get(mid) > target:
                    j = mid-1
                else:
                    return mid
            return -1
        
        def search2(i,j):
            while i<=j:
                mid = (i+j)//2
                if mountainArr.get(mid) < target:
                    j = mid-1
                elif mountainArr.get(mid) > target:
                    i = mid+1
                else:
                    return mid
            return -1

        
        i = 0
        j = mountainArr.length()-1

        while i<=j:
            mid = (i+j)//2
            left = mountainArr.get(mid-1)
            index = mountainArr.get(mid)
            right = mountainArr.get(mid+1)

            if left < index and index > right:
                a = search1(0,mid)
                b = search2(mid, mountainArr.length()-1)
                if a == -1 and b==-1:
                    return -1
                elif a!=-1:
                    return a
                else:
                    return b
            elif left < index:
                 i = mid+1
            else:
                j = mid-1

File: Data Structures & Algorithms/search-in-rotated-sorted-array-ii/submission-3.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int i=0, j=nums.size()-1;
        while(i<j){
            int mid = (i+j)/2;
            if(nums[mid] == target){
                return true;
            } else if(nums[mid] == nums[i]){
                i++;
            } else if(nums[mid] == nums[j]){
                j--;
            } else if(nums[mid] > nums[j]){
                i = mid+1;
            } else {
                j = mid;
            }
        }
        int k = i-1, l = i;
        cout<<i;
        j = nums.size()-1, i = 0;
        while(i<=k){
            int mid = (i+k)/2;
            if(nums[mid] == target){
                return true;
            }else if(nums[mid] > target){
                k = mid-1;
            } else {
                i = mid+1;
            }
        }

        while(l<=j){
            int mid = (l+j)/2;
            if(nums[mid] == target){
                return true;
            }else if(nums[mid] > target){
                j = mid-1;
            } else {
                l = mid+1;
            }
        }
        return false;
    }
};